### ISSUE TITLE
Is there a way to load a TF graph just once, and run it multiple times?

### ISSUE THREAD

**Issue by sekharvth** [tensorflow/tensorflow#19969]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: 

Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:

Windows 10 Home, running TF on Colab and on VM on GCP.

- **TensorFlow installed from (source or binary)**: 

Binary
- **TensorFlow version (use command below)**:

1.8.0
- **Python version**: 

3.6.5
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: Tesla K80, 12 GB
- **Exact command to reproduce**:

### Describe the problem

I have a TF model that was saved with the tf.train.Saver class, and so I have the .meta file, the data-00000-of-00001 file, the index file and the checkpoint file.

I use it for inference like this :

```
    graph_num = tf.Graph()
    with graph_num.as_default():
        sess = tf.Session()
        with sess.as_default():

            new_saver = tf.train.import_meta_graph('{}.meta'.format(model_path), clear_devices=True)
            new_saver.restore(sess, ('{}'.format(model_path)))
            sess.run(tf.tables_initializer())

            arr_placeholder = graph_num.get_operation_by_name('arr_placeholder/inp_array').outputs[0]
            str_placeholder = graph_num.get_operation_by_name('str_placeholder/inp_string').outputs[0]
            dropout_keep_prob = graph_num.get_operation_by_name('dropout_keep_prob/keep_prob').outputs[0]

            logis = graph_num.get_tensor_by_name('logits/preds/BiasAdd:0')

            def model_api(input_data):
                # ...preprocessing the input_data...

                a = sess.run(tf.nn.softmax(logis),
                             feed_dict={
                                 arr_placeholder:
                                     np.array(list_of_primary_inputs).reshape(len(list_of_primary_inputs), 142),
                                 dropout_keep_prob: 1.0, str_placeholder: input_string
                             })

                return a
```

So far so good, but then I call the function like this: 

```
tf.reset_default_graph()
result = model_api(test_input_data)
```
and each time I call it, it gives me different results for the same test data.

But when I instantiate the graph again and then call the function, it gives me the same numbers.

This behaviour is rather odd, and I don't want to re load the graph every time I want to pass in a new instance(s).

I can't use a for loop within the session, because the instances to be predicted come in real time, and so I have to use a function that supports arguments.

I saw this post too : https://stackoverflow.com/questions/42124000/reuse-graph-and-use-it-multiple-times but that wasn't helping my case.

I tried freezing the graph ( converting the existing meta graph into .pb ) but that too was giving me an error with one of the lookup tables that I have. And that is filed as a separate issue on GitHub, and unfortunately the workaround mentioned there didn't work for me. That issue is still open : https://github.com/tensorflow/tensorflow/issues/8665

I have even set tf.set_random_seed to a constant value while training, and tried doing the same to the inference part as well, but to no avail.

So right now I'm at my wits end.

I have asked this question on SO ( https://stackoverflow.com/questions/50812641/load-a-tensorflow-graph-once-and-use-it-multiple-times ) too, but that hasn't been helpful.

Why does it give me different results each time? And is there a way to load the graph once, and then keep running new instances without running into this issue of inconsistent outputs?

I couldn't find such a query while researching on this topic, so if there is currently no way to load a graph and run it multiple times all the while producing consistent outputs, can this be considered a feature request?

**Comment by krishngithub** [comment]
have requirement on similar lines, at the inference time would like to have same(fixed) graph which to be executed for different run time inputs. Currently graph is getting loaded for each request which is over head and taking 1 sec more for every request

for eg: we have below approx wrapper which will be called in by webserver via POST/GET at runtime.
```
service_wrapper(query):
   _input_fn =Lambda: _populatedataset_fn(query)
   return model.predict(_input_fn)
```
now we are looking for a way where we can expose this as a api in a cloud service which should avoid loading graph multiple times for each query request by the client

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by cy89** [comment]
@suharshs can you please comment on how to avoid graph reloading?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by sekharvth** [comment]
The random outputs problem has been solved, attributed to a bug in Google Sentence Encoder, which was part of the graph, corrected in a later release. ( https://github.com/tensorflow/hub/issues/74 )
But the reloading part is still a problem. I'm currently using sess.as_default(), and never explicitly closing the session, but this just keeps consuming memory on each POST request.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by sekharvth** [comment]
Issue solved after using multiple functions that pass around the session and tensors within a sess.as_default block.

**Comment by santhoshkolloju** [comment]
Hi ,

Can you please elaborate this solution. i have a similar problem while using REST API

**Comment by sekharvth** [comment]
What is the specific problem you're facing?

**Comment by coooold** [comment]
check this

`myGraph = tf.Graph()
sess = tf.Session(graph=myGraph)
with myGraph.as_default():
    with sess.as_default() as sess:
        trainStep()

sess.run(loss)`
